c3po/job/utils/metadata.py

```diff
@@ -0,0 +1,72 @@
+from c3po.db.dao import user, link, song, artist, genre
+from c3po.db.common.base import session_factory
+from datetime import datetime
+from music_metadata_extractor import SongData
+
+session = session_factory()
```

Comment:
I wonder this is the correct place to instantiate session. Because this would be initialized only once (when this module is imported). But we have multiple points where we have `session.commit()`. I have a feeling we might need to create a new session (which really isn't that expensive and our application doesn't need to be performant since it's a background session) each time. Can you confirm this behaviour?

Reply:
Insert operations are working with a single session. Creating a new session each time will most probably also work, I'll check.

Reply:
I tried and it doesn't seem to be working with creating a new session each time. It returns an error saying the object is being accessed in another session. we will probably have to create seperate functions for inserting `ArtistGenre` and `ArtistSong` that have independent sessions of their own to ensure that only one session is open at a time. Should we go ahead with that?

Reply:
[Looking at the SQLAlchemy docs](https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it), seems like having a single session seems to work because the transaction scope ends every time we commit, but the session scope continues to create a new transaction scope post the commit automatically. Buy my gripe with having an overarching session like this is that if we are accessing 2 different objects - and there is a failure in one of them, there is no sane way to rollback only those changes.

Furthermore, while not a problem currently (hopefully never), this works primarily because our application is single-threaded. It will break the moment we change that pattern.

So, I think the best place to create a session would be at the start of where we create a chain of interrelated transactions. So the `insert` method here is a great place (which again, makes the presence of a class weird - although not _wrong_ technically). So to conclude, what IMO would be a good way to fix what we have here is:
1. Get rid of the class, create a session at the start of the chain of functions (rather than at the module level) and pass it around.
2. Add a context manager ([like as shared in the official docs](https://stackoverflow.com/a/29805305/4396392)) in combination with `session_factory()` so we can handle failures gracefully.

This gives us a clear view of the session scope, the transactions scope, can handle failures cleanly and gets rid of the redundant class.

Reply:
I am a bit unfamiliar with using context manager. I'll try to get some code pushed.